### PR TITLE
CPM-907 : Allow create/update with Family Generation

### PR DESCRIPTION
**Pull request by Beninho**
**Description (for Contributor and Core Developer)**

Allow create/update with Family Generation

- Add validation for structure property
- Add tests

**Definition Of Done (for Core Developer only)**

- [X] Tests

### CODE REVIEW

File: components/identifier-generator/back/src/Domain/Model/Property/FamilyProperty.php [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+/**
+ * @copyright 2023 Akeneo SAS (https://www.akeneo.com)
+ * @license   https://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ *
+ * @phpstan-type FamilyPropertyNormalized array{type: 'family', process: ProcessNormalized}
+ */
+final class FamilyProperty implements PropertyInterface
+{
+    public function __construct(
```
Comment (pierallard) [inline components/identifier-generator/back/src/Domain/Model/Property/FamilyProperty.php:17]: I don't find any new FamilyProperty(...), please set this private
```suggestion
    private function __construct(
```

File: components/identifier-generator/back/src/Domain/Model/Property/Process.php [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+final class Process
+{
+    public const PROCESS_TYPE_NO = 'no';
+    public const PROCESS_TYPE_TRUNCATE = 'truncate';
+    public const PROCESS_TYPE_NOMENCLATURE = 'nomenclature';
+
+    public const PROCESS_OPERATOR_EQ = 'EQUALS';
+    public const PROCESS_OPERATOR_LTE = 'LOWER_OR_EQUAL_THAN';
+
+    public function __construct(
```
Comment (pierallard) [inline components/identifier-generator/back/src/Domain/Model/Property/Process.php:25]: Same, you have only 1 place in the code with new Process, and it's in a test ; you should set this private.
```suggestion
    private function __construct(
```

File: components/identifier-generator/back/src/Domain/Model/Property/Process.php [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+namespace Akeneo\Pim\Automation\IdentifierGenerator\Domain\Model\Property;
+
+use Webmozart\Assert\Assert;
+
+/**
+ * @copyright 2023 Akeneo SAS (https://www.akeneo.com)
+ * @license   https://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ *
+ * @phpstan-type ProcessOperator 'EQUALS'|'LOWER_OR_EQUAL_THAN'
+ * @phpstan-type ProcessNormalized array{type: 'no'|'truncate'|'nomenclature', operator?: string, value?: int}
```
Comment (pierallard) [inline components/identifier-generator/back/src/Domain/Model/Property/Process.php:14]: I think you can be more specific about typing, but not sure
Is it possible to write something like
```
array{type: 'no'}|array{type: 'truncate', operator: string, value: int}
```
?
As the card said, don't talk about nomenclature for now.

File: components/identifier-generator/back/src/Domain/Model/Property/Process.php [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+ */
+final class Process
+{
+    public const PROCESS_TYPE_NO = 'no';
+    public const PROCESS_TYPE_TRUNCATE = 'truncate';
+    public const PROCESS_TYPE_NOMENCLATURE = 'nomenclature';
```
Comment (pierallard) [inline components/identifier-generator/back/src/Domain/Model/Property/Process.php:20]: Remove this

File: components/identifier-generator/back/src/Domain/Model/Property/Process.php [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        ]);
+    }
+
+    /**
+     * @inerhitdoc
+     */
+    public static function fromNormalized(array $normalizedProperty): Process
+    {
+        Assert::keyExists($normalizedProperty, 'type');
+        Assert::oneOf($normalizedProperty['type'], [self::PROCESS_TYPE_NO, self::PROCESS_TYPE_TRUNCATE, self::PROCESS_TYPE_NOMENCLATURE]);
```
Comment (pierallard) [inline components/identifier-generator/back/src/Domain/Model/Property/Process.php:55]:
```suggestion
        Assert::oneOf($normalizedProperty['type'], [self::PROCESS_TYPE_NO, self::PROCESS_TYPE_TRUNCATE]);
```

File: components/identifier-generator/back/src/Domain/Model/Property/Process.php [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        if ('truncate' === $normalizedProperty['type']) {
+            Assert::keyExists($normalizedProperty, 'operator');
+            Assert::notEmpty($normalizedProperty['operator']);
+            Assert::oneOf($normalizedProperty['operator'], [self::PROCESS_OPERATOR_EQ, self::PROCESS_OPERATOR_LTE]);
+
+            Assert::keyExists($normalizedProperty, 'value');
+            Assert::numeric($normalizedProperty['value']);
```
Comment (pierallard) [inline components/identifier-generator/back/src/Domain/Model/Property/Process.php:62]: You have to assert value is > 0 (and I think there is a max too)

File: components/identifier-generator/back/src/Domain/Model/Property/Process.php [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+            $normalizedProperty['value'] = intval($normalizedProperty['value']);
+        }
+        $operator = array_key_exists('operator', $normalizedProperty) ? $normalizedProperty['operator'] : null;
+        $value = array_key_exists('value', $normalizedProperty) ? $normalizedProperty['value'] : null;
```
Comment (pierallard) [inline components/identifier-generator/back/src/Domain/Model/Property/Process.php:66]:
```suggestion
        $operator = $normalizedProperty['operator'] ?? null;
        $value = $normalizedProperty['value'] ?? null;
```

File: components/identifier-generator/back/src/Infrastructure/Symfony/Resources/config/validation/create_generator_command.yml [hunk 8]
```diff
@@ -59,6 +60,18 @@ Akeneo\Pim\Automation\IdentifierGenerator\Application\Create\CreateGeneratorComm
                             - Optional:
                                 - GreaterThanOrEqual: !php/const Akeneo\Pim\Automation\IdentifierGenerator\Domain\Model\Property\AutoNumber::LIMIT_DIGITS_MIN
                                 - LessThanOrEqual: !php/const Akeneo\Pim\Automation\IdentifierGenerator\Domain\Model\Property\AutoNumber::LIMIT_DIGITS_MAX
+                        process:
+                            - Optional:
+                                - Collection:
+                                    fields:
+                                        type:
+                                            - Choice:
+                                                choices: ['no', 'truncate', 'nomenclature']
```
Comment (pierallard) [inline components/identifier-generator/back/src/Infrastructure/Symfony/Resources/config/validation/create_generator_command.yml:69]:
```suggestion
                                                choices: ['no', 'truncate']
```

File: components/identifier-generator/back/src/Infrastructure/Symfony/Resources/config/validation/update_generator_command.yml [hunk 9]
```diff
@@ -58,6 +59,18 @@ Akeneo\Pim\Automation\IdentifierGenerator\Application\Update\UpdateGeneratorComm
                             - Optional:
                                 - GreaterThanOrEqual: !php/const Akeneo\Pim\Automation\IdentifierGenerator\Domain\Model\Property\AutoNumber::LIMIT_DIGITS_MIN
                                 - LessThanOrEqual: !php/const Akeneo\Pim\Automation\IdentifierGenerator\Domain\Model\Property\AutoNumber::LIMIT_DIGITS_MAX
+                        process:
+                            - Optional:
+                                - Collection:
+                                    fields:
+                                        type:
+                                            - Choice:
+                                                choices: ['no', 'truncate', 'nomenclature']
```
Comment (pierallard) [inline components/identifier-generator/back/src/Infrastructure/Symfony/Resources/config/validation/update_generator_command.yml:68]:
```suggestion
                                                choices: ['no', 'truncate']
```

File: components/identifier-generator/back/src/Infrastructure/Symfony/Resources/translations/validators.en_US.yml [hunk 10]
```diff
@@ -18,3 +18,6 @@ validation:
         unknown_scope: 'The "{{ scopeCode }}" scope does not exist.'
         unknown_locale: 'The "{{ localeCode }}" locale does not exist or is not activated.'
         inactive_locale: 'The "{{ localeCode }}" locale is not active for the "{{ scopeCode }}" channel.'
+        family_property_fields_required: '"{{field}}" field is required for "family" type.'
```
Comment (pierallard) [inline components/identifier-generator/back/src/Infrastructure/Symfony/Resources/translations/validators.en_US.yml:21]:
```suggestion
        family_property_fields_required: '"{{ field }}" field is required for "family" type.'
```

File: components/identifier-generator/back/src/Infrastructure/Validation/FamilyPropertyShouldBeValidValidator.php [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+            return;
+        }
+        if ($property['type'] !== FamilyProperty::type()) {
+            return;
+        }
+
+        if (!\array_key_exists('process', $property)) {
+            $this->context
+                ->buildViolation($constraint->fieldsRequired, [
+                    '{{field}}' => 'process',
```
Comment (pierallard) [inline components/identifier-generator/back/src/Infrastructure/Validation/FamilyPropertyShouldBeValidValidator.php:45]:
```suggestion
                    '{{ field }}' => 'process',
```

File: components/identifier-generator/back/src/Infrastructure/Validation/FamilyPropertyShouldBeValidValidator.php [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+
+                break;
+            case Process::PROCESS_TYPE_TRUNCATE:
+                $this->validateProcessTypeTruncate($property['process'], $constraint);
+
+                break;
+            case Process::PROCESS_TYPE_NOMENCLATURE:
+                $this->validateProcessTypeNomenclature($property['process'], $constraint);
+
+                break;
```
Comment (pierallard) [inline components/identifier-generator/back/src/Infrastructure/Validation/FamilyPropertyShouldBeValidValidator.php:68]:
```suggestion
```

File: components/identifier-generator/back/src/Infrastructure/Validation/FamilyPropertyShouldBeValidValidator.php [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+        }
+    }
+
+    private function validateProcessTypeNo(array $property, FamilyPropertyShouldBeValid $constraint): void
+    {
+        $this->validator->inContext($this->context)->validate($property, new Collection([
+            'fields' => [
+                'type' => null,
+            ],
+            'extraFieldsMessage' => $constraint->processTypeNoOtherProperties,
```
Comment (pierallard) [inline components/identifier-generator/back/src/Infrastructure/Validation/FamilyPropertyShouldBeValidValidator.php:78]: I think you should not put extraFieldsMessage here, and let Symfony manage this alone (and drop the message in the Constraint and in the validators.en_US.yml files)

File: components/identifier-generator/back/src/Infrastructure/Validation/FamilyPropertyShouldBeValidValidator.php [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+                        'max' => 5,
+                    ]),
+                ],
+            ],
+        ]));
+    }
+
+    private function validateProcessTypeNomenclature(array $property, FamilyPropertyShouldBeValid $constraint): void
+    {
+    }
```
Comment (pierallard) [inline components/identifier-generator/back/src/Infrastructure/Validation/FamilyPropertyShouldBeValidValidator.php:106]:
```suggestion
```

File: components/identifier-generator/back/tests/Specification/Domain/Model/Property/FamilyPropertySpec.php [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+    {
+        $this->normalize()->shouldReturn([
+            'type' => 'family',
+            'process' => [
+                'type' => 'truncate',
+                'operator' => 'EQUALS',
+                'value' => 3
+            ]
+        ]);
+    }
```
Comment (pierallard) [inline components/identifier-generator/back/tests/Specification/Domain/Model/Property/FamilyPropertySpec.php:54]: Please add a test for each Assert you set in the constructor

File: components/identifier-generator/back/tests/features/createIdentifierGenerator.feature [hunk 16]
```diff
@@ -98,6 +98,68 @@ Feature: Create Identifier Generator
...
+    And the identifier should not be created
+
+  Scenario: Cannot create an identifier generator with empty family property
+    When I try to create an identifier generator with empty family process property
+    Then I should get an error with message 'structure[0][process][type]: This field is missing.'
+    And the identifier should not be created
+
+  Scenario: Cannot create an identifier generator with family process type unknown
+    When I try to create an identifier generator with a family process with type unknown and operator undefined and undefined as value
+    Then I should get an error with message 'structure[0][process][type]: Type "unknown" can only be one of the following: "no", "truncate", "nomenclature".'
```
Comment (pierallard) [inline components/identifier-generator/back/tests/features/createIdentifierGenerator.feature:119]:
```suggestion
    Then I should get an error with message 'structure[0][process][type]: Type "unknown" can only be one of the following: "no", "truncate".'
```

File: components/identifier-generator/back/tests/features/createIdentifierGenerator.feature [hunk 17]
```diff
@@ -98,6 +98,68 @@ Feature: Create Identifier Generator
...
+    When I try to create an identifier generator with a family process with type truncate and operator EQUALS and bad as value
+    Then I should get an error with message 'structure[0][value]: This value should be of type digit.'
+    And the identifier should not be created
+
+  Scenario: Cannot create an identifier generator with a family containing truncate process and value not in range
+    When I try to create an identifier generator with a family process with type truncate and operator EQUALS and 0 as value
+    Then I should get an error with message 'structure[0][value]: This value should be between 1 and 5.'
+    And the identifier should not be created
+
+  Scenario: Can create an identifier generator with a no family process
```
Comment (pierallard) [inline components/identifier-generator/back/tests/features/createIdentifierGenerator.feature:153]: Same remark for edit feature
```suggestion
  Scenario: Can create an identifier generator with a family property and no process
```

File: components/identifier-generator/back/tests/features/createIdentifierGenerator.feature [hunk 18]
```diff
@@ -98,6 +98,68 @@ Feature: Create Identifier Generator
...
+    When I try to create an identifier generator with a family process with type no and operator undefined and undefined as value
+    Then The identifier generator is saved in the repository
+    And I should not get any error
+
+  Scenario: Can create an identifier generator with a truncate family process
```
Comment (pierallard) [inline components/identifier-generator/back/tests/features/createIdentifierGenerator.feature:158]: Same remark for edit feature
```suggestion
  Scenario: Can create an identifier generator with a family property and a truncate process
```

**Review by pierallard — COMMENTED** [review]
Noice

**Review by pierallard — APPROVED** [review]

**Review by cemorin — APPROVED** [review]

**Pull request merged**